grid/gpt/generate-text/maps/openai.suma

```diff
@@ -0,0 +1,180 @@
+profile = "gpt/generate-text@1.0"
+provider = "openai"
+
+map CompleteText { 
+  http POST "/v1/completions" {
+    security "bearer"
+  
+    request "application/json" {
+      headers {
+        "OpenAI-Organization" = parameters.organizationId
+      }
+
+      body {
+        prompt = input.prompt
+        model = call MapCompleteModelSize(model = input.model)
+        temperature = call MapCreativity(creativity = input.creativity)
+        max_tokens = call MapMaxLength(approxMaxWords = input.approxMaxWords)
+        n = call MapCount(count = input.count)
+        stream = false
+        logprobs = null
+      }
+    }
+
+    response 200 "application/json" {
+      completions = call foreach(choice of body.choices) MapCompletion(
+        choice = choice
+      )
+
+      map result {
+        completions = completions
+      }
+    }
+
+    response "application/json" {
+      error = call HandleErrorWithJson(body = body, statusCode = statusCode)
+      return map error error
+    }
+
+    response {
+      error = call HandleError(body = body, statusCode = statusCode)
+      return map error error
+    }
+  }
+}
+
+map EditText { 
+  http POST "/v1/edits" {
+    security "bearer"
+  
+    request "application/json" {
+      headers {
+        "OpenAI-Organization" = parameters.organizationId
+      }
+
+      body {
+        input = input.text
+        instruction = input.instructions
+        model = "text-davinci-edit-001"
+        temperature = call MapCreativity(creativity = input.creativity)
+        n = call MapCount(count = input.count)
+      }
+    }
+
+    response 200 "application/json" {
+      edits = call foreach(choice of body.choices) MapEdit(
+        choice = choice
+      )
+
+      map result {
+        edits = edits
+      }
+    }
+
+    response "application/json" {
+      error = call HandleErrorWithJson(body = body, statusCode = statusCode)
+      return map error error
+    }
+
+    response {
+      error = call HandleError(body = body, statusCode = statusCode)
+      return map error error
+    }
+  }
+}
+
+operation MapCompleteModelSize {
+  MODELSIZE_TO_OPENAI_MODEL = {
+    "small": "text-babbage-001",
+    "medium": "text-curie-001",
+    "large": "text-davinci-003"
+  }
+
+  DEFAULT_MODEL_SIZE = "large"
+
+  return MODELSIZE_TO_OPENAI_MODEL[args.model || DEFAULT_MODEL_SIZE]
+}
+
+operation MapCreativity {
+  DEFAULT_CREATIVITY = 1
+  
+  creativity = args.creativity || DEFAULT_CREATIVITY
+
+  // Clamp `creativity` between 0 and 1
+  return Math.min(Math.max(creativity, 0), 1)
+}
+
+operation MapMaxLength {
+  // About tokens: https://beta.openai.com/tokenizer
+
+  DEFAULT_MAX_WORDS = 12
+
+  return parseInt((args.approxMaxWords || DEFAULT_MAX_WORDS) * 1.4)
```

Comment:
question: So you are transforming words to OpenAI tokens. The 1.4 constant will work for English, what about other languages?

Reply:
> The 1.4 constant will work for English, what about other languages?

For English, the actual best ratio is somewhere around `1.2`. I tried Czech phrases and it was around `1.6`, so I thought 1.4 might be fine. For that reason, I also call the property in profile an _approximate_.

Do you find this a blocker for using?

Reply:
> Do you find this a blocker for using?

IMO for basic use cases this abstraction will be good enough.

On the other hand tokens are core of the GPT-3 technology. OpenAI monetisation is based on them. There is this OpenAI document describing how complex the topic is https://help.openai.com/en/articles/4936856-what-are-tokens-and-how-to-count-them. We are hiding the complexity by using inaccurate constant.

Reply:
The complexity is actually the reason for abstraction, though, no?

My thinking is:
— I don't know how tokens are calculated
— They are different for every language (which means I need to use different values for different languages anyway)
— I kind of still need to experiment with it to determine the optimal max tokens value

So to me, it feels like it's _“random”_ anyway, but the word abstraction makes it easier for me to at least get some idea (approximate) of what I should be expecting.

Good point on the monetization, though. However don't the points above apply as well?

Reply:
BTW Please don't take this as arguing your points. I'm trying to figure what's the best for the profile. Leaking the _tokens_ thing in there feels 🤢 but if it would be totally necessary, then we'll do it 🤷

Reply:
> The complexity is actually the reason for abstraction, though, no?
I agree.

Reply:
I see now. Thank you 🙏